File: app/src/main/java/com/tao/chinachuclient/Show_Image.java
package com.tao.chinachuclient;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;

import com.tenthbit.view.ZoomImageView;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.DialogInterface.OnClickListener;
import android.content.Intent;
import android.graphics.BitmapFactory;
import android.os.Bundle;
import android.os.Environment;
import android.support.v7.app.AppCompatActivity;
import android.util.Base64;
import android.view.View;
import android.widget.EditText;
import android.widget.Toast;

public class Show_Image extends AppCompatActivity{

	private String programId;
	private int pos;
	private byte[] byteImage;
	private ZoomImageView image;

	@Override
	protected void onCreate(Bundle savedInstanceState){
		super.onCreate(savedInstanceState);
		setContentView(R.layout.show_image);
		image = (ZoomImageView)findViewById(R.id.show_image_image);

		Intent i = getIntent();
		final String base64 = i.getStringExtra("base64");
		pos = i.getIntExtra("pos", -1);
		programId = i.getStringExtra("programId");
		byteImage = Base64.decode(base64, Base64.DEFAULT);

		image.setImageBitmap(BitmapFactory.decodeByteArray(byteImage, 0, byteImage.length));
	}

	public void image_option_click(View v){
		new AlertDialog.Builder(this)
		.setMessage("保存しますか？")
		.setNegativeButton("キャンセル", null)
		.setPositiveButton("OK", new OnClickListener(){
			@Override
			public void onClick(DialogInterface dialog, int which){
				saveImage();
			}
		}).show();
	}

	public void saveImage(){
		final String fileName;
		if(pos == -1)
			fileName = programId;
		else
			fileName = programId + "-" + pos;
		final String type = ".jpg";
		final String saveDir = Environment.getExternalStorageDirectory().getAbsolutePath() + "/" + Environment.DIRECTORY_DOWNLOADS;
		final String imgPath = saveDir + "/" + fileName + type;

		if(new File(imgPath).exists()) {
			new AlertDialog.Builder(this)
			.setTitle("エラー:ファイルが既に存在しています")
			.setItems(new String[]{"上書き", "ファイル名を指定して保存", "キャンセル"}, new OnClickListener(){
				@Override
				public void onClick(DialogInterface dialog, int which){
					if(which == 0) {
						save(imgPath);
					}else if(which == 1) {
						final EditText edit = new EditText(Show_Image.this);
						edit.setText(fileName);
						new AlertDialog.Builder(Show_Image.this)
						.setTitle("ファイル名を指定してください")
						.setView(edit)
						.setNegativeButton("キャンセル", null)
						.setPositiveButton("OK", new OnClickListener(){
							@Override
							public void onClick(DialogInterface dialog, int which){
								String newPath = saveDir + "/" + edit.getText().toString() + type;
								if(new File(newPath).exists())
									saveImage();
								else
									save(newPath);
							}
						}).show();
					}
				}
			}).show();
		}else{
			save(imgPath);
		}
	}

	public void save(String imgPath){
		FileOutputStream fos;
		try{
			fos = new FileOutputStream(imgPath, true);
			fos.write(byteImage);
			fos.close();
		}catch(IOException e){
			Toast.makeText(this, e.getMessage(), Toast.LENGTH_LONG).show();
			return;
		}
		Toast.makeText(this, "保存しました\n" + imgPath, Toast.LENGTH_LONG).show();
	}
}

File: app/src/main/java/com/tao/chinachuclient/ShowImage.kt
package com.tao.chinachuclient

import android.Manifest
import android.app.AlertDialog
import android.content.pm.PackageManager
import android.graphics.BitmapFactory
import android.os.Bundle
import android.os.Environment
import androidx.core.app.ActivityCompat
import androidx.core.content.PermissionChecker
import androidx.appcompat.app.AppCompatActivity
import android.util.Base64
import android.view.View
import android.widget.EditText
import android.widget.Toast
import com.tenthbit.view.ZoomImageView
import java.io.File
import java.io.FileOutputStream
import java.io.IOException

class ShowImage : AppCompatActivity() {

    companion object {
        const val REQUEST_CODE_WRITE_EXTERNAL_STORAGE = 810
    }

    private lateinit var programId: String
    private var pos: Int = -1
    private lateinit var byteImage: ByteArray

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.show_image)
        val image = findViewById<ZoomImageView>(R.id.showImageImage)

        val base64 = intent.getStringExtra("base64")
        pos = intent.getIntExtra("pos", -1)
        programId = intent.getStringExtra("programId")!!
        byteImage = Base64.decode(base64, Base64.DEFAULT)

        image.setImageBitmap(BitmapFactory.decodeByteArray(byteImage, 0, byteImage.size))
    }

    fun clickImageOption(@Suppress("UNUSED_PARAMETER") v: View) {
        AlertDialog.Builder(this)
                .setMessage(R.string.is_save)
                .setNegativeButton(R.string.cancel, null)
                .setPositiveButton(R.string.ok) { _, _ ->
                    saveImage()
                }
                .show()
    }

    private fun saveImage() {
        if (!hasWriteExternalStoragePermission()) {
            requestWriteExternalStoragePermission()
            return
        }

        val fileName = if (pos == -1) {
            programId
        } else {
            "$programId-$pos"
        }
        val type = ".jpg"
        val saveDir = Environment.getExternalStorageDirectory().absolutePath + "/" + Environment.DIRECTORY_DOWNLOADS
        val imgPath = "$saveDir/$fileName$type"

        if (File(imgPath).exists()) {
            AlertDialog.Builder(this)
                    .setTitle(R.string.error_file_already_exists)
                    .setItems(resources.getStringArray(R.array.file_already_exists_fix_suggestion)) { _, which ->
                        if (which == 0) {
                            save(imgPath)
                        } else if (which == 1) {
                            val edit = EditText(this)
                            edit.setText(fileName)
                            AlertDialog.Builder(this)
                                    .setTitle(R.string.assign_file_name)
                                    .setView(edit)
                                    .setNegativeButton(R.string.cancel, null)
                                    .setPositiveButton(R.string.ok) { _, _ ->
                                        val newPath = "$saveDir/${edit.text}$type"
                                        if (File(newPath).exists()) {
                                            saveImage()
                                        } else {
                                            save(newPath)
                                        }

                                    }.show()
                        }
                    }
                    .show()
        } else {
            save(imgPath)
        }
    }

    private fun save(imgPath: String) {
        try {
            val fos = FileOutputStream(imgPath, true)
            fos.write(byteImage)
            fos.close()
        } catch (e: IOException) {
            Toast.makeText(this, e.message, Toast.LENGTH_LONG).show()
            return
        }
        Toast.makeText(this, getString(R.string.saved) + "\n" + imgPath, Toast.LENGTH_LONG).show()
    }

    private fun hasWriteExternalStoragePermission(): Boolean {
        val writeExternalStorage = PermissionChecker.checkSelfPermission(this, Manifest.permission.WRITE_EXTERNAL_STORAGE)
        return (writeExternalStorage == PackageManager.PERMISSION_GRANTED)
    }

    private fun requestWriteExternalStoragePermission() {
        ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), REQUEST_CODE_WRITE_EXTERNAL_STORAGE)
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestCode != REQUEST_CODE_WRITE_EXTERNAL_STORAGE) {
            return
        }
        if (permissions[0] == Manifest.permission.WRITE_EXTERNAL_STORAGE && grantResults[0] == PackageManager.PERMISSION_GRANTED) {
            saveImage()
        } else {
            Toast.makeText(applicationContext, R.string.fail_permission, Toast.LENGTH_LONG).show()
        }
    }

}